examples/bri-3/src/bri/zkp/agents/proof.agent.ts

```diff
@@ -0,0 +1,49 @@
+import { BadRequestException, Injectable } from '@nestjs/common';
+import {
+  INVALID_PROOF_INPUT,
+  NOT_FOUND_ERR_MESSAGE,
+} from '../api/err.messages';
+import { Proof } from '../models/proof';
+import { ProofStorageAgent } from './proofStorage.agent';
+import { BpiAccount } from '../../identity/bpiAccounts/models/bpiAccount';
+
+@Injectable()
+export class ProofAgent {
+  constructor(private storageAgent: ProofStorageAgent) {}
+
+  public throwIfProofInputInvalid(document: string): boolean {
+    if (document == '') {
+      throw new BadRequestException(INVALID_PROOF_INPUT);
+    }
+    return true;
+  }
+
+  public createNewProof(
+    id: string,
+    owner: BpiAccount,
+    document: string,
+    signature: string,
+  ): Proof {
+    const payload =
+      this.convertDocumentToPayloadAndThrowIfDocumentValidationFails(document);
+
+    return new Proof(id, owner, payload, signature);
+  }
+
+  public async verifyDocumentWithProof(document: string): Promise<boolean> {
+    const proofToVerify =
+      this.convertDocumentToPayloadAndThrowIfDocumentValidationFails(document);
+    const verified = await this.storageAgent.verifyProofInShieldContract(
```

Comment:
again are we verifying the proof offchain or onchain or both? and let's please get rid of the word `shieldContract` and say `verifyProofOnchain` if onchain and `verifyProofOffchain` if not onchain

Reply:
Verifying the proof on-chain. Will remove `shieldContract`

Reply:
@biscuitdey Is verifying off-chain out of scope? We should be able to do both, should we not? Perhaps its best to label as out of scope for the time being given time constraints.

Reply:
we must do both

Reply:
During our last call, we agreed on the following points:

1) Content addressable hash of a document (root of the merkle tree representing the latest state of the document) is stored onchain only.  This hash can be used by the BPI to verify the latest state of the document as well as workgroup participants if they want to verify that their  system of records are in sync outside of the BPI.

2) Additional feature which will not be covered as part of this PR is that the history of the state changes of the document will be stored on chain as well, so that the interested party can verify  state transitions of the document independently of the BPI. 

3) More complex ZKP schemes are out of scope of SRI work and will be tackled after this grant is completed.

For the current PR, main changes are to apply point 1 and to rename proof to anchorHash everywhere (anyone has a better name proposal?) and to rename ShieldContract to OnChain.

@biscuitdey please have a look and let us know if all is clear. I suggest we do another pass on the PR once the changes are made.

@Therecanbeonlyone1969 @skosito @Kasshern @Manik-Jain please review this comment and add if i missed anything.